pkg/secrets/secretsdb.go

```diff
@@ -0,0 +1,192 @@
+package secrets
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+type db struct {
+	// Secrets maps a secret id to secret metadata
+	Secrets map[string]Secret `json:"secrets"`
+	// NameToID maps a secret name to a secret id
+	NameToID map[string]string `json:"nametoid"`
+	// IDToName maps a secret id to a secret name
+	IDToName map[string]string `json:"idtoname"`
+	// lastModified is the time when the database was last modified on the file system
+	lastModified time.Time
+}
+
+// loadDB loads database data into the in-memory cache if it has been modified
+func (s *SecretsManager) loadDB() error {
+	// check if the db file exists
+	fileInfo, err := os.Stat(s.secretsDBPath)
+	if err != nil {
+		if !os.IsExist(err) {
+			// if the file doesn't exist, then there's no reason to update the db cache
+			// the db cache will show no entries anyway
+			// the file will be created later on a store()
+			return nil
+		} else {
+			return err
+		}
+	}
+
+	// we check if the file has been modified after the last time it was loaded into the cache
+	// if the file has been modified, then we know that our cache is not up-to-date, so we load
+	// the db into the cache
+	if !s.db.lastModified.Equal(fileInfo.ModTime()) {
+		file, err := os.Open(s.secretsDBPath)
+		if err != nil {
+			return err
+		}
+		defer file.Close()
+		if err != nil {
+			return err
+		}
+
+		byteValue, err := ioutil.ReadAll(file)
+		if err != nil {
+			return err
+		}
+		unmarshalled := new(db)
+		if err := json.Unmarshal(byteValue, unmarshalled); err != nil {
+			return err
+		}
+		s.db = unmarshalled
+		s.db.lastModified = fileInfo.ModTime()
+	}
+	return nil
+}
+
+// getNameAndID takes a secret's name or id and returns both its
+// identifying name and id
+func (s *SecretsManager) getNameAndID(nameOrID string) (name, id string, err error) {
+	err = s.loadDB()
+	if err != nil {
+		return "", "", err
+	}
+	if id, ok := s.db.NameToID[nameOrID]; ok {
+		name := nameOrID
+		return name, id, nil
+	}
+
+	if name, ok := s.db.IDToName[nameOrID]; ok {
+		id := nameOrID
+		return name, id, nil
+	}
+
+	// ID prefix may have been given, iterate through all ID's
+	exists := false
+	var foundID, foundName string
+	for id, name := range s.db.IDToName {
+		if exists {
+			return "", "", errors.Wrapf(ErrAmbiguous, "more than one result secret with prefix %s", nameOrID)
```

Comment:
Are we testing only for looking up the _last_ secret?

Reply:
No, we're seeing if the prefix is unique: that the given ID prefix matches only one full ID. 
ie:
secret 1: id `aaaaaaaaaaaaaaaaaaaaaaaaa` 
secret 2: id `abcdefghijklmnopqrstuvwxyz` 

searching for secret `aaa` would match secret 1, but searching for secret `a` would fail with `ErrAmbiguous`

The boolean `exists` before the `HasPrefix(id, nameOrID)` checks if the prefix has already matched, and returns with an error. I tested it it seemed to work correctly?

Reply:
Let's assume we hit line 91 and set `exists = true`.  Wouldn't the next iteration immediately error out (independent of the prefix)?

I think we need to move the if branch from line 87 to line 91.  Only there we know that `nameOrID` is a prefix of `id`.

Or am I missing something?  My brain isn't up to speed yet after PTO

Reply:
Nope, you're right, I'm on PTO brain :)